### PR TITLE
More thiserror

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Use of `core::Result` creates circular dependencies.

#### Summary of Changes

* Building on #7177, this PR explores what the codebase would look like if `thiserror` was used more widely.
* `core::Result` changed to a private module

### CODE REVIEW

**Comment by garious** [comment]
Anyone care to review? Nothing too controversial here, I think. If anything, it's that the same error, such as an RecvTimeoutError, can be nested in multiple top-level errors, so you can't check for "any RecvTimeoutError".  I don't see that being problematic, but putting the same error in 2 places allows for a stupid bug if your code checks for that error in the wrong place.

File: core/src/poh_recorder.rs [hunk 1]
```diff
@@ -27,17 +26,28 @@ use std::cmp;
...
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Error, Debug, Clone)]
 pub enum PohRecorderError {
+    #[error("invalid calling object")]
     InvalidCallingObject,
+
+    #[error("max height reached")]
     MaxHeightReached,
+
+    #[error("max height not reached")]
```
Comment (CriesofCarrots) [inline core/src/poh_recorder.rs]:
```suggestion
    #[error("min height not reached")]
```

File: core/src/commitment.rs [hunk 2]
```diff
@@ -113,15 +112,10 @@ impl AggregateCommitmentService {
...
-                        if let Err(e) = Self::run(&receiver, &block_commitment_cache, &exit_) {
-                            match e {
-                                Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
-                                Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
-                                _ => info!(
-                                    "Unexpected error from AggregateCommitmentService: {:?}",
-                                    e
-                                ),
-                            }
+                        if let Err(RecvTimeoutError::Disconnected) =
```
Comment (CriesofCarrots) [inline core/src/commitment.rs]: Any particular reason you removed the `match` here and in the transaction_status_service, but not in similar constructions below?

Comment (garious) [reply core/src/commitment.rs]: I only removed the match when the new error enum was so small that there was no longer any unexpected error cases.

File: core/src/poh_recorder.rs (hunk 1 above)
Comment (garious) [reply core/src/poh_recorder.rs]: Thanks! Fixed in 81d1e85c95b6187ecd50c849fb19a365dd2d1b28

**Pull request merged**